lib/bundle/defaults.js

```diff
@@ -19,75 +19,163 @@ function getGenericDefaults (generator) {
+  }
+  if (canReferenceResponse(pathParsed)) {
+    return "#/responses";
+  }
+
+  return "#/definitions";
+};
+
+const defaultOas3RootResolver = (pathFromRoot) => {
+  const pathParsed = pathFromRoot.split("/");
+
+  if (canReferenceParameter(pathParsed)) {
+    return "#/components/parameters";
+  }
+  if (canReferenceResponse(pathParsed)) {
+    return "#/components/responses";
+  }
+  if (canReferenceRequestBody(pathParsed)) {
+    return "#/components/requestBodies";
+  }
+  if (canReferenceHeader(pathParsed)) {
+    return "#/components/headers";
+  }
+
+  return "#/components/schemas";
+};
+
+module.exports.defaultOas2RootResolver = defaultOas2RootResolver;
+module.exports.defaultOas3RootResolver = defaultOas3RootResolver;
+
+module.exports.getDefaultsForOAS2 = function (defaults = getGenericDefaults(new KeyGenerator(module.exports.defaultOas2RootResolver))) {
   return {
     ...defaults,
     generateKey (schema, file, hash, pathFromRoot) {
-      pathFromRoot = normalizeOasSchemasHash(pathFromRoot, defaults.defaultRoot);
-
-      if (!pathFromRoot.startsWith(defaults.defaultRoot) && !isSchemaPlacement(pathFromRoot.split("/"))) {
+      const pathParsed = pathFromRoot.split("/");
+
+      if (
+        !pathFromRoot.startsWith(defaults.defaultRoot(pathFromRoot)) &&
+        !canReferenceSchema(pathParsed) &&
+        !canReferenceParameter(pathParsed) &&
+        !canReferenceResponse(pathParsed)
+      ) {
         return null;
       }
 
       if (hash !== "#" && hash !== null) {
-        return defaults.generateKey(schema, file, normalizeOasSchemasHash(hash, defaults.defaultRoot), pathFromRoot);
+        return defaults.generateKey(schema, file, normalizeOasSchemasHash(hash, defaults.defaultRoot(pathFromRoot)), pathFromRoot);
       }
 
       return defaults.generateKey(schema, file, hash, pathFromRoot);
     },
   };
 };
 
-module.exports.getDefaultsForOAS3 = function (defaults = getGenericDefaults(new KeyGenerator("#/components/schemas"))) {
-  return module.exports.getDefaultsForOAS2(defaults);
+module.exports.getDefaultsForOAS3 = function (defaults = getGenericDefaults(new KeyGenerator(defaultOas3RootResolver))) {
+  return {
+    ...defaults,
+    generateKey (schema, file, hash, pathFromRoot) {
+      const pathParsed = pathFromRoot.split("/");
+
+      if (
+        !pathFromRoot.startsWith(defaults.defaultRoot(pathFromRoot)) &&
+        !canReferenceSchema(pathParsed) &&
+        !canReferenceParameter(pathParsed) &&
+        !canReferenceResponse(pathParsed) &&
+        !canReferenceRequestBody(pathParsed) &&
+        !canReferenceHeader(pathParsed)
+      ) {
+        return null;
+      }
+
+      if (hash !== "#" && hash !== null) {
+        return defaults.generateKey(schema, file, normalizeOasSchemasHash(hash, defaults.defaultRoot(pathFromRoot)), pathFromRoot);
+      }
+
+      return defaults.generateKey(schema, file, hash, pathFromRoot);
+    },
+  };
 };
 
 function normalizeOasSchemasHash (hash, root) {
```

Comment:
Let's remove this function altogether.
The idea behind that function was to detect cross spec references, but this was totally optional.
By cross spec referencing, I mean referencing a schema from an oas2 doc inside of an oas3 doc.
The world won't end if we don't support it - in fact it's not quite valid to do.

Reply:
@P0lip  If I remove it, I get errors like:

<img width="529" alt="Screen Shot 2021-02-15 at 11 42 15" src="https://user-images.githubusercontent.com/1868852/107936472-ff9f6580-6f82-11eb-95c4-c98a5e64d7ea.png">

I'm not sure how to then resolve it?

Reply:
Yeah, that's fine. As said, this is a mixed oas2<->oas3 referencing.
It's kind of like you did the following

```json
"components": {
  "parameters": {
    "foo": { "$ref:" blabla.json#/parameters/foo"
  }
}
```

So, we can either get rid of entirely, or support all components.

Reply:
Ok, I'll update the tests then 👍